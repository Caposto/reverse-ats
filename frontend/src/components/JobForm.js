import React, {useState} from 'react';
import { Button, Form, Input } from 'semantic-ui-react';

export const JobForm = () => {
    const [description, setDescription] = useState(''); // Create state for job description, default empty string

    // Return form for inputting job description
    // TODO: Add <Button> from semantic-ui for submitting the form (In Ben Awad's Video 15:00/)
    return (
        <Form> 
            <Form.Field>
                <Input 
                  placeholder="job description" 
                  value={description} 
                  onChange={e => setDescription(e.target.value)}  
                />
            </Form.Field>
            <Button onClick={async() => {
                const job_description = {description};
                const response = await fetch('/extract_keywords', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(job_description)
                })

                if (response.ok) {
                    let res = response.json()
                    console.log(res);
                }
            }}>Submit</Button>
        </Form>
    )
}