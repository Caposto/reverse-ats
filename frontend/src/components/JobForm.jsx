import { useState, useContext } from "react";
import PropTypes from "prop-types";
import KeywordList from "./KeywordList";
import Error from "./Error";
import getKeywordsFromFlask from "../services/User";
import KeywordsContext from "../services/KeywordContext";

function JobForm({ descriptionType }) {
  const [description, setDescription] = useState("");
  const [visible, setVisibility] = useState(true);
  const [loading, setLoading] = useState(false);
  const [keywords, setKeywords] = useState([]);
  const [count, setCount] = useState(0);
  const [error, setError] = useState(false);
  const [exception, setException] = useState(false);
  const [submitted, setSubmitted] = useState(false);

  const { handleJobKeywords, handleResumeKeywords } = useContext(KeywordsContext);

  const maxJobDescriptionLength = 5000;

  const handleSubmit = async (e) => {
    e.preventDefault();
    // Handle empty string and whitespace submissions
    if (description.trim().length === 0) {
      setException(true);
      setDescription("");
      setCount(0);
    } else {
      try {
        setException(false);
        setLoading(true);

        // Ensure consitency of keyword list when comparing
        const apiKeywords = await getKeywordsFromFlask(description);
        const formattedKeywords = apiKeywords.map((keyword, index) => ({
          id: index + 1,
          text: keyword,
          isComplete: false,
        }));

        setKeywords(apiKeywords);

        // Update parent keywords with useContext
        if (descriptionType === "job") {
          handleJobKeywords(formattedKeywords);
        } else if (descriptionType === "resume") {
          handleResumeKeywords(formattedKeywords);
        }

        setDescription("");
        setCount(0);
        setVisibility(false);
        setLoading(false);
        setSubmitted(true);
      } catch (err) {
        setLoading(false);
        setError(true);
      }
    }
  };

  const submitNewDescription = () => {
    setKeywords([]);
    setVisibility(true);
    setSubmitted(false);
    // updateParentKeywords([]);
  };

  return (
    <div aria-label="form-test-id" className="px-8">
      {!submitted && (
        <form onSubmit={handleSubmit}>
          <textarea
            aria-label="Job Description Submission"
            placeholder="Enter Text"
            value={description}
            onChange={(e) => {
              setDescription(e.target.value);
              setCount(e.target.value.length);
            }}
            rows="6"
            maxLength={maxJobDescriptionLength}
            className="block px-4 w-full rounded-md border-0 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:py-1.5 sm:text-sm sm:leading-6"
          />
          <div className="pt-2">
            Character Limit: {count}/{maxJobDescriptionLength}
          </div>
          <div className="py-4">
            <button
              className="text-xl p-2 rounded-md border border-2"
              type="submit"
              aria-label="Submit Button"
            >
              Extract
            </button>
          </div>
        </form>
      )}
      <div className="text-xl"> {loading ? <>Loading...</> : ""} </div>
      {!visible && !error && !exception && (
        <div>
          <KeywordList initial={keywords} />
          <button
            className="text-xl p-2 rounded-md border border-2"
            type="submit"
            onClick={submitNewDescription} // FIXME: Is there a way to reload just the component instead of whole page?
          >
            Submit New Description
          </button>
        </div>
      )}
      {error && <Error />}
      {exception && (
        <Error errorMessage="No empty submissions allowed! Please try again." />
      )}
    </div>
  );
}

JobForm.propTypes = {
  descriptionType: PropTypes.string.isRequired,
};

export default JobForm;
