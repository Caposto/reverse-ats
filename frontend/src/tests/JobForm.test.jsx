import renderer from "react-test-renderer";
import JobForm from "../components/JobForm";

it("changes the class when hovered", () => {
  const component = renderer.create(<JobForm />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseEnter();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseLeave();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
