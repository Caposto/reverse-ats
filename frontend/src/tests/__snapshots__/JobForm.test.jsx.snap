// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `
<div
  className="p-8"
>
  <form
    onSubmit={[Function]}
  >
    <textarea
      className="block px-4 w-full rounded-md border-0 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:py-1.5 sm:text-sm sm:leading-6"
      maxLength={5000}
      onChange={[Function]}
      placeholder="Enter Job Description"
      rows="6"
      value=""
    />
    <div
      className="pt-2"
    >
      Character Limit: 
      0
      /
      5000
    </div>
    <div
      className="py-4"
    >
      <button
        className="text-xl p-2 rounded-md border border-2"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
